crates/iceberg/src/expr/visitors/inclusive_projection.rs

```diff
@@ -0,0 +1,371 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+use crate::expr::visitors::bound_predicate_visitor::{visit, BoundPredicateVisitor};
+use crate::expr::{BoundPredicate, BoundReference, Predicate};
+use crate::spec::{Datum, PartitionField, PartitionSpecRef};
+use crate::Error;
+use fnv::FnvHashSet;
+use std::collections::HashMap;
+use std::ops::Not;
+
+pub(crate) struct InclusiveProjection {
+    partition_spec: PartitionSpecRef,
+    cached_parts: HashMap<i32, Vec<PartitionField>>,
+}
+
+impl InclusiveProjection {
+    pub(crate) fn new(partition_spec: PartitionSpecRef) -> Self {
+        Self {
+            partition_spec,
+            cached_parts: HashMap::new(),
+        }
+    }
+
+    fn get_parts_for_field_id(&mut self, field_id: i32) -> &Vec<PartitionField> {
+        if let std::collections::hash_map::Entry::Vacant(e) = self.cached_parts.entry(field_id) {
+            let mut parts: Vec<PartitionField> = vec![];
+            for partition_spec_field in &self.partition_spec.fields {
+                if partition_spec_field.source_id == field_id {
+                    parts.push(partition_spec_field.clone())
+                }
+            }
+
+            e.insert(parts);
+        }
+
+        &self.cached_parts[&field_id]
+    }
+
+    pub(crate) fn project(&mut self, predicate: &BoundPredicate) -> crate::Result<Predicate> {
```

Comment:
I've taken a closer look and I think we're missing a thing here?

I believe we have to rewrite the Predicate with `rewrite_not`,
since it's assumed the ExprTree does not contain any NOT nodes.
https://github.com/apache/iceberg-python/blob/main/pyiceberg/expressions/visitors.py#L805-L810

However, this implies that the `InclusiveProjection` operates on `Predicate` rather than `BoundPredicate`. This means we have to not only apply `rewrite_not` but also `bind` before we call `visit` here.

Reply:
As per comment below, `rewrite_not` should have already been applied by the time we attempt an `InclusiveProjection`. I've updated to `panic` if a Not operator is encountered during projection.

Reply:
If you do it that way, you invoke `rewrite-not` every time that you call `project`, which could be several times per invocation of `plan_files` in a table scan, and `plan_files` itself can get invoked any number of times on a particular instance of a `TableScan`. In the proposed design `rewrite-not` only ever gets called once, when the `TableScan` gets built, guaranteeing that the predicate is not-free already by the time it gets to the `InclusiveProjection`.